### PR TITLE
Rename Serializer to Output and move to `flash.core.data.io.output`

### LINKED ISSUE #917: Rename `Serializer` to `Output`
And rename `serialize` method to `transform` or similar.

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Fixes #917 

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/general/predictions.rst [hunk 1]
```diff
@@ -71,7 +71,7 @@ reference below).
     # 2. Load the model from a checkpoint
     model = ImageClassifier.load_from_checkpoint("https://flash-weights.s3.amazonaws.com/image_classification_model.pt")
 
-    # 3. Attach the Serializer
+    # 3. Attach the Output
     model.serializer = Probabilities()
```
Comment (tchaton) [inline docs/source/general/predictions.rst:75]: Should it be model.output = ProbabilitiesOuptut ?

File: docs/source/template/optional.rst [hunk 2]
```diff
@@ -25,16 +25,16 @@ Add output serializers to your Task
 
 We recommend that you do most of the heavy lifting in the :class:`~flash.core.data.process.Postprocess`.
 Specifically, it should include any formatting and transforms that should always be applied to the predictions.
-If you want to support different use cases that require different prediction formats, you should add some :class:`~flash.core.data.process.Serializer` implementations in a ``serialization.py`` file.
+If you want to support different use cases that require different prediction formats, you should add some :class:`~flash.core.data.io.output.Output` implementations in a ``serialization.py`` file.
```
Comment (tchaton) [inline docs/source/template/optional.rst:28]: Should we say output.py ?

File: flash/core/model.py [hunk 3]
```diff
@@ -624,20 +618,16 @@ def deserializer(self, deserializer: Union[Deserializer, Mapping[str, Deserializ
...
-        return self._serializer
+    def output(self) -> Optional[Output]:
+        """The current :class:`.Output` associated with this model."""
+        return self._output
 
     @torch.jit.unused
-    @serializer.setter
-    def serializer(self, serializer: Union[Serializer, Mapping[str, Serializer]]):
+    @output.setter
+    def output(self, serializer: Union[Output, Mapping[str, Output]]):
```
Comment (tchaton) [inline flash/core/model.py:627]:
```suggestion
    def output(self, output: Union[Output, Mapping[str, Output]]):
```

File: flash/core/model.py [hunk 4]
```diff
@@ -624,20 +618,16 @@ def deserializer(self, deserializer: Union[Deserializer, Mapping[str, Deserializ
...
         if isinstance(serializer, Mapping):
-            serializer = SerializerMapping(serializer)
-        self._serializer = serializer
+            serializer = OutputMapping(serializer)
```
Comment (tchaton) [inline flash/core/model.py:629]: Let's use output instead no ?

File: flash/template/classification/model.py [hunk 5]
```diff
@@ -67,7 +67,7 @@ def __init__(
             metrics=metrics,
             learning_rate=learning_rate,
             multi_label=multi_label,
-            serializer=serializer or Labels(),
+            output=output or Labels(),
```
Comment (tchaton) [inline flash/template/classification/model.py:70]: Personal preference, but I would prefer `LabelsOutput` as we would have `InputImagePaths`. This would make it more consistent and readable too.

File: docs/source/template/optional.rst (hunk 2 above)
Comment (ethanwharris) [reply docs/source/template/optional.rst]: Yeah, done :smiley:

File: flash/core/model.py (hunk 3 above)
Comment (ethanwharris) [reply flash/core/model.py]: Done

File: flash/template/classification/model.py (hunk 5 above)
Comment (ethanwharris) [reply flash/template/classification/model.py:70]: I think we should leave that for a potential follow up as we would need to deprecate `Labels` etc. rather than just rename as they are a public API.

**Review by ananyahjha93 — APPROVED** [review]

File: flash/image/segmentation/serialization.py [hunk 6]
```diff
@@ -90,7 +90,7 @@ def _visualize(self, labels):
         plt.imshow(labels_vis)
         plt.show()
 
-    def serialize(self, sample: Dict[str, torch.Tensor]) -> torch.Tensor:
+    def transform(self, sample: Dict[str, torch.Tensor]) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (ethanwharris) [reply flash/image/segmentation/serialization.py:93]: Maybe, although it definitely shouldn't be called serialize any more. As the role of `Output` is to convert raw predictions into the desired output format I think transform is fine. But it could otherwise be something more explicit like `convert(raw_predictions)` rather than `transform(sample)`

File: docs/source/general/predictions.rst [hunk 7]
```diff
@@ -77,8 +77,8 @@ reference below).
         "https://flash-weights.s3.amazonaws.com/0.5.2/image_classification_model.pt"
     )
 
-    # 3. Attach the Serializer
-    model.serializer = Probabilities()
+    # 3. Attach the Output
+    model.output = Probabilities()
```
Comment (tchaton) [inline docs/source/general/predictions.rst:81]: Should we call them ProbabilitiesOutput ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: flash/image/segmentation/serialization.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply flash/image/segmentation/serialization.py:93]: I was thinking about `export_transform`.
